feat: support more types in switch statements

mattjohnsonpint opened the pull request:
Fixes #2919
Fixes #648

Changes proposed in this pull request:

- Fixes compiler error when using a `switch` statement with certain types.
  - All numeric types that were already supported
  - Add support for strings, floats, and 64-bit ints
- Adds tests to validate

I've reused the existing equality comparison logic, so any behavior for `==` is retained for switch cases, including overloads, etc.

- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

HerrCai0907 reviewed (changes requested):
as does not use string pool..so it is impossible to compare two string directly with ptr. string.eq must be used for each check.

mattjohnsonpint commented:
> as does not use string pool..so it is impossible to compare two string directly with ptr. string.eq must be used for each check.

... oh.  I see.  I was thrown off because the naive test passed, but this one fails:

```ts
assert(doSwitchString("o" + "n" + "e") == 1);
```

because the concatenated resulting string is in a different variable and thus a different pointer than the `"one"` in the switch case.

I'll add a special case for strings. I think it should be good for the rest though.

CountBleck commented:
I wonder if this should just be using the `==` operator overload for all objects that have them...

mattjohnsonpint commented:
Working on a revision.  I think I can reuse `makeEq`

mattjohnsonpint commented:
This is now updated with the revisions.  I've split the existing `compileCommutativeCompareBinaryExpression` method such that the core logic can be re-used when evaluating the switch cases.  I've also added more tests to show that it works as expected.  Thanks.

mattjohnsonpint commented:
@HerrCai0907 any other adjustments you would like?
